swift/Sources/SignalClient/SealedSender.swift

```diff
@@ -0,0 +1,164 @@
+
+    public func publicKey() throws -> PublicKey {
+        return try invokeFnReturningPublicKey {
+            signal_server_certificate_get_key($0, nativeHandle)
+        }
+    }
+}
+
+
+public class SenderCertificate: ClonableHandleOwner {
+    public init<Bytes: ContiguousBytes>(_ bytes: Bytes) throws {
+        let handle: OpaquePointer? = try bytes.withUnsafeBytes {
+            var result: OpaquePointer?
+            try checkError(signal_sender_certificate_deserialize(&result, $0.baseAddress?.assumingMemoryBound(to: UInt8.self), $0.count))
+            return result
+        }
+        super.init(owned: handle!)
+    }
+
+    internal override class func destroyNativeHandle(_ handle: OpaquePointer) {
+        signal_sender_certificate_destroy(handle)
+    }
+
+    public func expiration() throws -> UInt64 {
+        return try invokeFnReturningInteger {
+            signal_sender_certificate_get_expiration(nativeHandle, $0)
+        }
+    }
+
+    public func deviceId() throws -> UInt32 {
+        return try invokeFnReturningInteger {
+            signal_sender_certificate_get_device_id(nativeHandle, $0)
+        }
+    }
+
+    public func serialize() throws -> [UInt8] {
+        return try invokeFnReturningArray {
+            signal_sender_certificate_get_serialized(nativeHandle, $0, $1)
+        }
+    }
+
+    public func certificateBytes() throws -> [UInt8] {
+        return try invokeFnReturningArray {
+            signal_sender_certificate_get_certificate(nativeHandle, $0, $1)
+        }
+    }
+
+    public func signatureBytes() throws -> [UInt8] {
+        return try invokeFnReturningArray {
+            signal_sender_certificate_get_signature(nativeHandle, $0, $1)
+        }
+    }
+
+    public func publicKey() throws -> PublicKey {
+        return try invokeFnReturningPublicKey {
+            signal_sender_certificate_get_key($0, nativeHandle)
+        }
+    }
+
+    public func senderUuid() throws -> String? {
+        return try invokeFnReturningOptionalString {
+            signal_sender_certificate_get_sender_uuid(nativeHandle, $0)
+        }
+    }
+
+    public func senderE164() throws -> String? {
+        return try invokeFnReturningOptionalString {
+            signal_sender_certificate_get_sender_e164(nativeHandle, $0)
+        }
+    }
+
+    public func serverCertificate() throws -> ServerCertificate {
+        var handle: OpaquePointer?
+        try checkError(signal_sender_certificate_get_server_certificate(&handle, nativeHandle))
+        return ServerCertificate(owned: handle!)
+    }
+
+    public func validate(trust_root: PublicKey, time: UInt64) throws -> Bool {
+        var result: UInt32 = 0
+        try checkError(signal_sender_certificate_validate(&result, nativeHandle, trust_root.nativeHandle, time))
+
+        if result == 1 {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    // FIXME missing signal_sender_certificate_preferred_address (is it needed?)
```

Comment:
The current logic I have is always preferring a UUID if present. Something like the preferred address logic may or may not make sense based on whether or not ProtocolAddress can be an opaque key.

Reply:
Preferring UUID definitely makes sense to me. Possibly this logic is also wrong for Android, or at the least should be left client-specific in some way.